src/support/threads.cpp

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2016 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <assert.h>
+
+#include <iostream>
+
+#include "threads.h"
+
+
+// debugging tools
+
+#ifdef BINARYEN_THREAD_DEBUG
+static std::mutex debug;
+#define DEBUG_PRINT(x) { std::lock_guard<std::mutex> lock(debug); x; }
+#else
+#define DEBUG_PRINT(x)
+#endif
+
+
+namespace wasm {
+
+// Global thread information
+
+std::atomic_bool setMainThreadId(false);
+std::thread::id mainThreadId;
+
+struct MainThreadNoter {
+  MainThreadNoter() {
+    // global ctors are called on main thread
```

Comment:
They're not :-)

http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html#2046
> The thread-id for global static initialization is unspecified, but it is not the default-constructed thread.

They can be called in parallel, or another thread than main.

Also: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0250r1.html